doc: add documentation with example for clipping planes (#1219)
cognitedata/reveal: documentation/docs/examples/clipping.mdx

---
id: clipping
title: Clipping/slicing planes
describe: Reveal support adding a number of slicing planes to "cut through" the models.
---

import { DemoWrapper } from '@site/src/components/DemoWrapper';
import useBaseUrl from '@docusaurus/useBaseUrl';

Slicing planes, also called clipping planes, can be used to "cut through" the model to reveal the interior. This is 
useful to isolate interesting parts of the model and hide obscuring geometry. Reveal supports having multiple slicing planes
in arbitrary position and orientation. When slicing planes are active,
the amount of geometry loaded might be reduced since culled geometry doesn't need
to be loaded (or rather loaded in a lower detail level).

<video preload="auto" autoPlay="autoplay" loop controls>
  <source src={useBaseUrl('/img/cad-clipping.webm')} />
</video>

> :::warning

Reviewer: There is no warning https://v2.docusaurus.io/docs/markdown-features#calloutsadmonitions

Perhaps it should be:
```suggestion
:::caution
```

Author: Interestingly, warning actually works (but caution is better here)

Clipping planes are currently not supported by point cloud models.
:::

## Examples

<DemoWrapper name="Cognite3DViewerDemo" />

### Single, fixed slicing plane

Slicing planes are defined by providing zero or more [`THREE.Plane`](https://threejs.org/docs/#api/en/math/Plane) 
to `Cognite3DViewer.setSlicingPlanes`. The slicing planes are activated for all models and
replaces any active slicing planes.

Typically, [setFromNormalAndCoplanarPoint](https://threejs.org/docs/#api/en/math/Plane.setFromNormalAndCoplanarPoint)
is used to create slicing planes. This function accepts a direction that defines the orientation of

> the plane and a point on the plane. For example will the following
> code create a slicing plane that only accepts geometry below `Y = 55`:

Reviewer: ```suggestion
the plane and a point on the plane. For example, the following
code will create a slicing plane that only accepts geometry below `Y = 55`:
```

```jsx runnable
const orientation = new THREE.Vector3(0, -1, 0);
const point = new THREE.Vector3(0, 55, 0);
const plane = new THREE.Plane().setFromNormalAndCoplanarPoint(orientation, point);
viewer.setSlicingPlanes([plane]);
```

### Animated slicing plane

> Planes can be manipulated runtime and effects will have effect in the viewer. Note however that rendering must be 
> triggered after updating the plane before the change will have visual effect. This can be done using 
> `Cognite3DViewer.forceRerender()`.

Reviewer: ```suggestion
Planes can be manipulated runtime and effects will have an effect on the viewer. Note however that rendering must be
triggered after updating the plane before the change will have a visual effect. This can be done using 
`Cognite3DViewer.forceRerender()`.
```

Author: I _think_ this is ok actually. If you're sure I'm wrong, please scream.

Below is an example animates a slicing plane by ping-ponging the plane in the vertical direction.

```js runnable
const bbox = model.getModelBoundingBox();

const orientation = new THREE.Vector3(0, -1, 0);
const point = new THREE.Vector3(0, bbox.max.y, 0);
const plane = new THREE.Plane();
viewer.setSlicingPlanes([plane]);

const start = performance.now();
setInterval(() => {
  // Ping-pong plane between top of model and half-way to the bottom
  const dt = (performance.now() - start) / 1000.0
  const s = 0.5 * (Math.sin(Math.PI * dt / 10.0) + 1.0);
  point.y = bbox.max.y - (bbox.max.y - bbox.min.y) / 2.0 * s;
  plane.setFromNormalAndCoplanarPoint(orientation, point);

  viewer.forceRerender();
}, 1000 / 60);
```
:::tip

> Updating `THREE.Plane`-instances will not trigger data loading. Typically you will need to trigger data loading if

Reviewer: I read this text multiple times to understand it, perhaps we just need some proper example here? I mean, the one where you do things right as you described here:

> Typically you will need to trigger data loading if you do substantial changes to the slicing planes.

Because currently the example looks cool, but it leaves me without knowledge on how to make things right. And it requires some patience to understand what is written in that tip, but basically, the tip just says - "hey, the example looks cool, but don't do that in real code, do X instead". So maybe we can have the X as an additional example here?

Reviewer: Just realized that "the X" example is the first one after checking the docs preview 🙄

Author: Updated a bit - hopefully more clear now :)

you do substantial changes to the slicing planes. In these cases, use `Cognite3DViewer.setSlicingPlanes`. The example 
above performs substantial changes to the planes without loading data. In a real application it's recommended to 
not load data during animation, but initiate data after an animation is completed.
:::

> ### CAD geometry filter

Reviewer: unexpectedly ended here (I know I draft thought, but anyway)

viewer.addObject3D(draggerMesh);

// Prepare a slicing plane
const orientation = new THREE.Vector3(0, -1, 0);
const plane = new THREE.Plane();
plane.setFromNormalAndCoplanarPoint(orientation, initialDragPosition);
viewer.setSlicingPlanes([plane]);

// Handle drag events
const controls = new DragControls([draggerMesh], viewer.getCamera(), viewer.domElement);
controls.addEventListener('drag', () => {
  // Fix XZ-poistion
  draggerMesh.position.set(initialDragPosition.x, draggerMesh.position.y, initialDragPosition.z);
  // Update slicing plane
  plane.setFromNormalAndCoplanarPoint(new THREE.Vector3(0, -1, 0), draggerMesh.position);
  viewer.forceRerender();
});
controls.addEventListener('dragstart', () => {
  draggerMaterial.color = new THREE.Color('gray');
  viewer.cameraControlsEnabled = false;
});
controls.addEventListener('dragend', () => {
  draggerMaterial.color = new THREE.Color('white');
  viewer.cameraControlsEnabled = true;
});
```

### CAD geometry filter

A special type of slicing is to provide a "geometry filter" when loading the model. A geometry filter restricts the 
area to load geometry for and is specified when the model is created.

> ```js 

Reviewer: is it intentionally non-runnable?

Author: Yes, because it must be applied on creation which we don't support now.
